Mathlib/Topology/UniformSpace/CoveringProfinite.lean

```diff
@@ -0,0 +1,115 @@
+  have step1 (x) : ∃ U : Opens X, x ∈ U ∧ ∀ y ∈ U, ∀ z ∈ U, (f y, f z) ∈ S := by
+    obtain ⟨R, hR, hR', hRS⟩ := comp_symm_of_uniformity hS
+    obtain ⟨U, hUB, hUo, hUx⟩ := mem_nhds_iff.mp <|  (f.continuousAt x).preimage_mem_nhds
+      <| UniformSpace.ball_mem_nhds _ hR
+    exact ⟨⟨U, hUo⟩, hUx, fun y hy z hz ↦ hRS <| prodMk_mem_compRel (hR' <| hUB hy) (hUB hz)⟩
+  choose U hUx hUS using step1
+  have hUc : IsOpenCover U := by ext x; simpa using ⟨x, hUx x⟩
+  -- Now refine it to a disjoint covering.
+  obtain ⟨n, W, hW₁, hW₂, hW₃⟩ := hUc.exists_finite_nonempty_disjoint_clopen_cover
+  refine ⟨n, W, fun j ↦ (hW₁ j).1, fun j y hy z hz ↦ ?_, hW₂, hW₃⟩
+  exact match (hW₁ j).2 with | ⟨i, hi⟩ => hUS i y (hi hy) z (hi hz)
+
+/--
+Any continuous function from a profinite space to a uniform space can be uniformly approximated
+by a function factoring through `Fin n`, for some `n`.
+-/
+lemma exists_fin_comp_of_mem_uniformity (hS : S ∈ 𝓤 V) :
+    ∃ (n : ℕ) (g : X → Fin n) (h : Fin n → V), Continuous g ∧ ∀ x, (f x, h (g x)) ∈ S := by
+  classical
+  obtain ⟨n, E, hEne, hES, hEuniv, hEdis⟩ :=
+    exists_disjoint_nonempty_clopen_cover_of_mem_uniformity f hS
+  have h_uniq (x) : ∃! i, x ∈ E i := by
+    refine match Set.mem_iUnion.mp (hEuniv <| mem_univ x) with
+      | ⟨i, hi⟩ => ⟨i, hi, fun j hj ↦ hEdis.eq ?_⟩
+    simpa [← Clopens.coe_disjoint, not_disjoint_iff] using ⟨x, hj, hi⟩
+  choose g hg hg' using h_uniq -- for each `x`, `g x` is the unique `i` such that `x ∈ E i`
+  have h_ex (i) : ∃ x, x ∈ E i := by
+    simpa [← SetLike.coe_set_eq, ← nonempty_iff_ne_empty] using hEne i
+  choose r hr using h_ex -- for each `i`, choose an `r i ∈ E i`
+  refine ⟨n, g, f ∘ r, continuous_discrete_rng.mpr fun j ↦ ?_, fun x ↦ (hES _) _ (hg _) _ (hr _)⟩
+  convert (E j).isOpen
+  exact Set.ext fun x ↦ ⟨fun hj ↦ hj ▸ hg x, fun hx ↦ (hg' _ _ hx).symm⟩
+
+/--
+If `f` is a continuous map from a profinite space to a uniform space with a group structure, then
+we can uniformly approximate `f` by finite products of indicator functions of clopen sets.
+-/
+@[to_additive "If `f` is a continuous map from a profinite space to a uniform space with an
+additive group structure, then we can uniformly approximate `f` by finite sums of indicator
+functions of clopen sets."]
+lemma exists_sum_const_mulIndicator_approx [CommGroup V] (hS : S ∈ 𝓤 V) :
+    ∃ (n : ℕ) (U : Fin n → Clopens X) (v : Fin n → V),
+    ∀ x, (f x, ∏ n, mulIndicator (U n) (fun _ ↦ v n) x) ∈ S := by
+  obtain ⟨n, g, h, hg, hgh⟩ := exists_fin_comp_of_mem_uniformity f hS
+  refine ⟨n, fun i ↦ ⟨_, (isClopen_discrete {i}).preimage hg⟩, h, fun x ↦ ?_⟩
+  convert hgh x
+  exact (Fintype.prod_eq_single _ fun i hi ↦ mulIndicator_of_not_mem hi.symm _).trans
+    (mulIndicator_of_mem rfl _)
+
+variable {R Y : Type*} [TopologicalSpace Y] [CompactSpace Y]
+  [TopologicalSpace R] [MonoidWithZero R]
+
+/-- A continuous function on `X × Y` can be uniformly approximated by sums of functions of the
```

Comment:
Perhaps you can specify in this doc the new setting (with the `R`-action). And what is the rationale behiind the choice of having some properties of `R` as variables and some only in the assumption of the lemma below? Finally, unlike the previous result, here `V` needs to be a `UniformAddGroup`, with a compatibility between the uniform structure and the group operation.

Reply:
The variables are split up because some, but not all, hypotheses are common to the two lemmas in this section. However I was playing a trick on the reader by using `V` for the ring in the latter lemma. I have adjusted this.